File: src/main/栈/t316/RemoveDuplicateLetters.java
package 栈.t316;

import java.util.LinkedList;

/**
 * @auther: sotowang
 * @date: 2020/2/13 19:47
 */
public class RemoveDuplicateLetters {
    public String removeDuplicateLetters(String s) {
        LinkedList<Character> stack = new LinkedList<>();
        char[] cs = s.toCharArray();
        for(int i=0;i<cs.length;i++){
            if(stack.contains(cs[i])){
                continue;
            }
            while(!stack.isEmpty() && stack.peek()>cs[i] && s.indexOf(stack.peek(),i) !=-1){
                stack.pop();
            }
            stack.push(cs[i]);
        }
        StringBuilder sb = new StringBuilder();
        for(int i=stack.size()-1;i>=0;i--){
            sb.append(stack.get(i));
        }
        return sb.toString();
    }
    public String removeDuplicateLetters1(String s) {
        if (s.length() < 2) {
            return s;
        }
        int[] dic = new int[26];
        boolean[] isPushed = new boolean[26];
        for (int i = 0; i < s.length(); i++) {
            dic[s.charAt(i) - 'a']++;
        }
        LinkedList<Character> stack = new LinkedList<>();
        stack.push(s.charAt(0));
        dic[s.charAt(0) - 'a']--;
        isPushed[s.charAt(0) - 'a'] = true;
        //cbacdcbc
        for (int i = 1; i < s.length(); i++) {
            char topChar = stack.peek();
            if (s.charAt(i) < topChar && !isPushed[s.charAt(i) - 'a']) {
                while (dic[topChar - 'a'] > 0 && s.charAt(i) <topChar) {
                    isPushed[stack.pop() - 'a'] = false;
                    if (stack.isEmpty()) {
                        break;
                    }
                    topChar = stack.peek();
                }
                stack.push(s.charAt(i));
            } else if (s.charAt(i) > topChar && !isPushed[s.charAt(i) - 'a']) {
                stack.push(s.charAt(i));
            }
            dic[s.charAt(i) - 'a']--;
            isPushed[s.charAt(i) - 'a'] = true;

        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.insert(0, stack.poll());
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        assert new RemoveDuplicateLetters().removeDuplicateLetters("bcabc").equals("abc");
        assert new RemoveDuplicateLetters().removeDuplicateLetters("cbacdbc").equals("acdb");

    }
}

File: src/main/栈/t316/removeDuplicateLetters.go
package main

func removeDuplicateLetters(s string) string {
	var stack []byte
	n := len(s)
	contians := func(stack []byte, key byte) bool {
		for _, v := range stack {
			if v == key {
				return true
			}
		}
		return false
	}
	indexOf := func(s string, fromIndex int, key byte) int {
		n := len(s)
		for i := fromIndex; i < n; i++ {
			if s[i] == key {
				return i
			}
		}
		return -1
	}
	for i := 0; i < n; i++ {
		if contians(stack, s[i]) {
			continue
		}
		for slen := len(stack); slen > 0 && stack[slen-1] > s[i] && indexOf(s, i, stack[slen-1]) != -1; {
			stack = stack[:slen-1]
			slen = len(stack)
		}
		stack = append(stack, s[i])
	}
	res := ""
	for _, v := range stack {
		res += string(v)
	}
	return res

}
func main() {
	s := "cbacdcbc"
	removeDuplicateLetters(s)
}
